presto-main/src/main/java/com/facebook/presto/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -336,7 +336,7 @@ else if (pendingSplits.isEmpty()) {
             return new ScheduleResult(false, overallNewTasks.build(), overallSplitAssignmentCount);
         }
 
-        if (anyBlockedOnPlacements || groupedExecution) {
```

Comment:
Do you want to elaborate about this change a bit? Is it because grouped execution is using `FixedSourcePartitionedScheduler`? I didn't see how this is related to `TODO` added below.

Reply:
Yes. There are three major responsibilities for `SourcePartitionedScheduler:
- Split schedule
- Task schedule
- Lifespan schedule

Note `SourcePartitionedScheduler` can be used either as a stage scheduler or source scheduler. In both case it needs to take care about split schedule. 

- When it's used as a stage scheduler, it's always used for *unpartitioned* stage (i.e. JOIN/AGGREGATE over *remote* source. 
       * It needs to take care about task schedule
       * It DOESN'T need to take care about lifespan schedule

- When it's used as a source scheduler, it's used by a `FixedSourcePartitionedScheduler`.
       * It DOESN'T need to take care about task schedule (although today it still do so) 
       * It need to take care about lifespan schedule (behemoth warning!)

Historically, other stage schedulers (notably `FixedSourcePartitionedScheduler`) doesn't call `stage.transitionToSchedulingSplits()`. And it  relies on `SourcePartitionedScheduler` to do this.

Now we make stage schedulers to call  `stage.transitionToSchedulingSplits()`. This frees `SourcePartitionedScheduler` to do this when it's used as a source scheduler. It only need to do this when it's used as a stage scheduler.

When grouped execution is enabled, `SourcePartitionedScheduler` is always used as a source scheduler. However, it's also possible `SourcePartitionedScheduler` is used as a source scheduler for ungrouped execution; and in that case `transitionToSchedulingSplits` will be done twice. And that's the TODO comment about :)

Reply:
Update the main comment as well: https://github.com/prestodb/presto/pull/12840#issue-280527871